### PR TITLE
Add a LAMMPS shell tool as alternative to the LAMMPS executable for interactive use

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

This PR implements a "LAMMPS shell", an alternative to the regular LAMMPS executable that is enhanced for interactive use with command line editing, command history, context aware TAB-expansion and more.

This can be used for learning LAMMPS by trying commands interactively (it won't abort when there is an error, so one does not have to start over) and can be used to embed a LAMMPS command  prompt into GUI applications (the GUI won't crash
if LAMMPS crashes).

To facilitate some of the enhancements additional introspection APIs had to be added to the library interface.

**Author(s)**

Axel Kohlmeyer, Temple U.

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No known issues.

**Implementation Notes**

The original idea was to have a LAMMPS executable that could be associated with LAMMPS input files in a desktop environment, but it has grown a bit from there.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated

### CODE REVIEW

**Review by sjplimp — APPROVED** [review]
ccol!

**Comment by sjplimp** [comment]
rather, cool!

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -642,6 +643,22 @@ if(BUILD_TOOLS)
   install(FILES ${LAMMPS_DOC_DIR}/msi2lmp.1 DESTINATION ${CMAKE_INSTALL_MANDIR}/man1)
 endif()
 
+if(BUILD_LAMMPS_SHELL)
+  include(FindPkgConfig)
```
Comment (junghans) [inline cmake/CMakeLists.txt:647]: I would use `find_package(PkgConfig)` here, more consistent  with the rest of the code.

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -642,6 +643,22 @@ if(BUILD_TOOLS)
...
+  pkg_check_modules(READLINE IMPORTED_TARGET readline)
+  if(PKG_CONFIG_FOUND AND READLINE_FOUND)
+    if(NOT LAMMPS_EXCEPTIONS)
+      message(WARNING "The LAMMPS shell needs LAMMPS_EXCEPTIONS enabled for full functionality")
+    endif()
+    add_executable(lammps-shell ${LAMMPS_TOOLS_DIR}/lammps-shell/lammps-shell.cpp)
+    target_compile_definitions(lammps-shell PRIVATE -DLAMMPS_LIB_NO_MPI)
+    target_link_libraries(lammps-shell PRIVATE lammps;PkgConfig::READLINE)
```
Comment (junghans) [inline cmake/CMakeLists.txt:655]:
```suggestion
    target_link_libraries(lammps-shell PRIVATE lammps PkgConfig::READLINE)
```

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -642,6 +643,22 @@ if(BUILD_TOOLS)
...
+    install(TARGETS lammps-shell DESTINATION ${CMAKE_INSTALL_BINDIR})
```
Comment (junghans) [inline cmake/CMakeLists.txt:656]:
```suggestion
    install(TARGETS lammps-shell EXPORT LAMMPS_Targets DESTINATION ${CMAKE_INSTALL_BINDIR})
```
export the target as well.

File: cmake/CMakeLists.txt [hunk 4]
```diff
@@ -642,6 +643,22 @@ if(BUILD_TOOLS)
...
+  else()
+    message(FATAL_ERROR "pkg-config and readline library not found. Cannot build LAMMPS shell")
```
Comment (junghans) [inline cmake/CMakeLists.txt:658]:
```suggestion
    message(FATAL_ERROR "pkg-config and/or readline library not found. Cannot build LAMMPS shell")
```

**Review by junghans — APPROVED** [review]
Yeah, it is super useful for tutorials! Great work.

Minor nitpicks in the code.

One question, why do we need a separate lammps executable for that? Most other tools just run the interactive mode when no input file or input on stdin is provided.

**Comment by akohlmey** [comment]
> One question, why do we need a separate lammps executable for that? Most other tools just run the interactive mode when no input file or input on stdin is provided.

There are multiple reasons:
- it has a different behavior when using the '-in/-i' command line flag. The normal LAMMPS executable exits at EOF, this drops you to a prompt. This is convenient for interactive use, since you "preload" a session with some that you don't have to type. It **can** detect non-interactive use (e.g. when using redirect) to have some compatibility with legacy tutorials that use redirect. for that TAB-expansion also needs to be disabled as input with TABs won't work.
- it has additional commands that are not LAMMPS commands, but "shell enhancements". there may be more coming. e.g. for use with a GUI. one idea would be to have an image viewer connected to a "dump image" command to have a view of the system as it is being manipulated.
- it won't work with multiple MPI ranks in interactive mode (hangs). multi-partition is even more complex...
- exception handling is a "soft" condition (to get better integration testing exposure), but in practice it is a "hard" condition and that is problematic with MPI parallel runs as has been discussed multiple times before.
- it is deliberately build on top of the library interface (with a few convenience functions from utils thrown in) so it is possible to integrate the code into building a more complex GUI application that can be built with only a few headers (library.h, lmptype.h and utils.h should suffice. **no** mpi.h) and not the full LAMMPS environment.

**Comment by arn-all** [comment]
This is a very nice tool ! Thanks for this improvement.
Some additional info I thought I'd share : 
- the package `readline-devel` (dnf - Fedora/CentOS) or `readline-dev` (apt - Ubuntu/Debian) should be installed.  Otherwise the build fails with `--   Package 'readline', required by 'virtual:world', not found CMake Error at /usr/share/cmake/Modules/FindPkgConfig.cmake:497 (message): A required package was not found`
- The command to build the shell tool is `cmake -DBUILD_LAMMPS_SHELL=yes -DLAMMPS_EXCEPTIONS=yes ../cmake`. The instruction in the README file "`setting the CMake variable BUILD_LAMMPS_SHELL to "on"`" (instead of "yes") might be misleading for unexperienced users.

**Comment by akohlmey** [comment]
> This is a very nice tool ! Thanks for this improvement.
> Some additional info I thought I'd share :
> 
> * the package `readline-devel` (dnf - Fedora/CentOS) or `readline-dev` (apt - Ubuntu/Debian) should be installed.  Otherwise the build fails with `--   Package 'readline', required by 'virtual:world', not found CMake Error at /usr/share/cmake/Modules/FindPkgConfig.cmake:497 (message): A required package was not found`

I'll review the README file and the documentation in the manual. This is all very new and was done rather quickly. I already noticed some formatting issues in the Manual that need to be corrected.

> * The command to build the shell tool is `cmake -DBUILD_LAMMPS_SHELL=yes -DLAMMPS_EXCEPTIONS=yes ../cmake`. The instruction in the README file "`setting the CMake variable BUILD_LAMMPS_SHELL to "on"`" (instead of "yes") might be misleading for unexperienced users.

Why is this misleading? Booleans in CMake can be "yes/no", "on/off", "1/0" etc. and we have been mostly using "on/off" throughout the LAMMPS manual, and **especially** in the CMake tutorial (which is what a person inexperienced with CMake should be reading) we explain this and try to use the "on/off" notation consistently: https://docs.lammps.org/Howto_cmake.html

Is there any obvious feature that is sill missing or some "readline power hint" that we could add to the manual?
So far, I have the following:
- add an explicit "save_history" command that allows to selectively write a range of commands from the history to a file (and not just the automatic history). and thus would allow to interactively build an input and then save and edit it (i.e. take out the mistakes)
- make "source" and alias for "include" as a more "natural" name for a command to read a file into a shell (while "include" is more fitting for a file).
- add some tips about customizing the .inputrc file: a) disable writing the .lammps_history file automatically, b) turning on tilde-expansion on filenames (so that hitting tab on e.g. "~/lammps/pote" will expand to "/home/<username>/lammps/potentials"); by default, it will complete on the name but leave the tilde, which will then fail since LAMMPS won't expand it.

**Comment by arn-all** [comment]
I am sorry, I was not aware that the yes/no and on/off were treated equivalently. 
I saw the yes/no notation in the build [[1]](https://docs.lammps.org/Build_basics.html) [[2]](https://docs.lammps.org/Build_settings.html) man pages, eg `-D BUILD_MPI=value        # yes or no, default is yes if CMake finds MPI, else no`. But as it's equivalent it's absolutely fine. Thanks for clarifying !

**Pull request merged**